text/0032-improved-api-tokens.md

```diff
@@ -0,0 +1,163 @@
+
+More details about Github's secret scanning program can be found [here](https://docs.github.com/en/developers/overview/secret-scanning-partner-program).
+
+At a high-level, we will need to be able to:
+
+- Receive a list of one or more potential tokens
+- [Verify the signature of the request](https://docs.github.com/en/developers/overview/secret-scanning-partner-program#implement-signature-verification-in-your-secret-alert-service)
+- Determine if the token is a true positive or a false positive
+  - send this feedback back to Github
+- Disable/revoke the compromised token
+- Send a notification to the appropriate user
+
+> :memo: This secret verification service is not an immediate need or a dependency of implementing the improved API tokens.
+
+# Background
+
+Accidental credential leaks [happen](https://github.com/ChALkeR/notes/blob/master/Do-not-underestimate-credentials-leaks.md). Even though we provide several tools for users to limit the storage of sensitive information in Sentry it can still happen.
+
+Our support for various authentication methods (including 2FA, SSO, and SAML) helps mitigate access of potentially sensitive information, but API tokens cannot support these additional authentication gates.
+
+We can help customers further protect their accounts and data by providing a means of auto-detecting leaked API tokens.
+
+# Options Considered
+
+## Option #1
+
+The migration from the legacy tokens to the new tokens should be rolled out over X releases. This allows for a smooth transition of self-hosted instances keeping pace with the versions.
+
+> With `n` being the current version.
+
+### `n+1`
+
+- Add a Django migration including the `hashed_token`, `name`, and `token_version` fields.
+  - columns should allow `null` values for now.
+- Remove displaying the plaintext token in the frontend, and only display newly created tokens once.
+- Implement new logic to generate the new token pattern.
+  - The `token_version` should be set as `2`.
+- Implement a backwards compatible means of verifying legacy tokens. This should:
+  - Verify the legacy token is valid/invalid.
+  - If valid, hash the legacy token and store it in the table.
+  - Remove the plaintext representation of the token in the table.
+  - Mark the `token_version` as `1`.
+- A notification/warning in the UI should be displayed recommending users recreate their tokens, resulting in the new token version.
+
+The `n+1` Sentry version should run in production for sufficient time to update the majority of the rows in the database as the legacy tokens are _used_ by users.
+
+### `n+2`
+
+- Add a Django migration to generate the `hashed_token` value for any tokens in `sentry_apitoken` that do not already have a hashed value.
+  - Rows matching this are legacy tokens that were not used during the transition period.
+  - The `token_version` should be set to `1` for these rows.
```

Comment:
I don't think we should try to straddle multiple token versions for a long period of time. If we update all the currently active tokens incrementally could we follow that up with a bulk operation to replace all the remaining tokens?

Reply:
+1

This is the goal. I'll update the proposal to clarify, but in general we would:

1. Release version 22.11.0 with the incremental token updates as they are used.
2. Release version 22.12.0 with the bulk operation to update any tokens that were not updated between the release dates.